File: apis/api-java/src/main/java/de/niklas/apijava/service/OrderService.java
package de.niklas.apijava.service;

import de.niklas.apijava.dto.order.*;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * OrderService kapselt die Logik zur Aggregation von Bestellungen
 */
@Service
public class OrderService {

    /**
     * Aggregiert die bezahlten Bestellungen pro Kunde
     * <p>
     * Für jeden Kunden werden Anzahl, Gesamtsumme sowie durchschnittlicher Betrag der Bestellungen berechnet
     * </p>
     * @param orders Eingabebestellungen (nur {@link STATUSENUM#PAID} werden berücksichtigt)
     * @return Aggregation pro Kunden als {@link OrderAggregateResponseDTO}
     */
    public OrderAggregateResponseDTO aggregateOrders(OrderRequestDTO[] orders) {
        Map<String, OrderResponseDTO> map = new HashMap<>();

        for (OrderRequestDTO order : orders) {
            if (order.status() != STATUSENUM.PAID) continue;

            OrderResponseDTO entry = map.get(order.customerId());
            if (entry == null) {
                entry = new OrderResponseDTO(order.customerId(), 0,0,0);
            }
            int newTotal = entry.totalOrders() +1;
            double newAmount = entry.amount() + order.amount();
            double newAvg = newAmount / newTotal;

            OrderResponseDTO updated = new OrderResponseDTO(order.customerId(), newTotal, newAmount, newAvg);
            map.put(order.customerId(), updated);
        }
        List<OrderResponseDTO> output = new ArrayList<>(map.values());
        output.sort((a, b) -> a.customerId().compareTo(b.customerId()));
        OrderResponseDTO[] customerArray = output.toArray(new OrderResponseDTO[0]);
        return new OrderAggregateResponseDTO(customerArray);
    }
}

File: apis/api-go/internal/service/order_service.go
package service

import (
	"apis/api-go/internal/dto"
	"sort"
)

// OrderService kapselt die Logik für die Order-Endpunkte
type OrderService struct{}

// NewOrderService erstellt eine neue OrderService-Instanz
func NewOrderService() *OrderService {
	return &OrderService{}
}

// Aggregate aggregiert bezahlte Bestellungen pro Kunde
func (os *OrderService) Aggregate(orders []dto.OrderInput) ([]dto.CustomerOut, error) {
	m := make(map[string]dto.CustomerOut, len(orders))

	for _, order := range orders {
		if order.Status != "PAID" {
			continue
		}
		entry, ok := m[order.CustomerId]
		if !ok {
			entry = dto.CustomerOut{
				CustomerId:  order.CustomerId,
				TotalOrders: 0,
				Amount:      0,
				Avg:         0,
			}
			m[order.CustomerId] = entry
		}
		entry.TotalOrders++
		entry.Amount += order.Amount
		entry.Avg = entry.Amount / float64(entry.TotalOrders)

		m[order.CustomerId] = entry
	}

	result := make([]dto.CustomerOut, 0, len(m))
	for _, v := range m {
		result = append(result, v)
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].CustomerId < result[j].CustomerId
	})

	return result, nil
}
